// @signalwire/node/dist/common/src/relay/calling/results/index.js
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
__export(require("./HangupResult"));
__export(require("./RecordResult"));
__export(require("./AnswerResult"));
__export(require("./PlayResult"));
__export(require("./PromptResult"));
__export(require("./ConnectResult"));
__export(require("./DialResult"));
__export(require("./FaxResult"));
__export(require("./DetectResult"));
__export(require("./TapResult"));
__export(require("./SendDigitsResult"));
__export(require("./DisconnectResult"));
class StopResult {
    constructor(result) {
        this.result = result;
        this.code = result.code;
        this.message = result.message;
        this.successful = result.code === '200';
    }
}
exports.StopResult = StopResult;
class PlayPauseResult {
    constructor(successful) {
        this.successful = successful;
    }
}
exports.PlayPauseResult = PlayPauseResult;
class PlayResumeResult {
    constructor(successful) {
        this.successful = successful;
    }
}
exports.PlayResumeResult = PlayResumeResult;
class PlayVolumeResult {
    constructor(successful) {
        this.successful = successful;
    }
}
exports.PlayVolumeResult = PlayVolumeResult;
class PromptVolumeResult {
    constructor(successful) {
        this.successful = successful;
    }
}
exports.PromptVolumeResult = PromptVolumeResult;
