// @signalwire/node/dist/common/src/messages/Verto.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const BaseRequest_1 = __importDefault(require("./verto/BaseRequest"));
const Login_1 = require("./verto/Login");
exports.Login = Login_1.Login;
const Result_1 = require("./verto/Result");
exports.Result = Result_1.Result;
const constants_1 = require("../webrtc/constants");
class Invite extends BaseRequest_1.default {
    toString() {
        return constants_1.VertoMethod.Invite;
    }
}
exports.Invite = Invite;
class Answer extends BaseRequest_1.default {
    toString() {
        return constants_1.VertoMethod.Answer;
    }
}
exports.Answer = Answer;
class Attach extends BaseRequest_1.default {
    toString() {
        return constants_1.VertoMethod.Attach;
    }
}
exports.Attach = Attach;
class Bye extends BaseRequest_1.default {
    toString() {
        return constants_1.VertoMethod.Bye;
    }
}
exports.Bye = Bye;
class Modify extends BaseRequest_1.default {
    toString() {
        return constants_1.VertoMethod.Modify;
    }
}
exports.Modify = Modify;
class Info extends BaseRequest_1.default {
    toString() {
        return constants_1.VertoMethod.Info;
    }
}
exports.Info = Info;
class Broadcast extends BaseRequest_1.default {
    toString() {
        return constants_1.VertoMethod.Broadcast;
    }
}
exports.Broadcast = Broadcast;
class Subscribe extends BaseRequest_1.default {
    toString() {
        return constants_1.VertoMethod.Subscribe;
    }
}
exports.Subscribe = Subscribe;
class Unsubscribe extends BaseRequest_1.default {
    toString() {
        return constants_1.VertoMethod.Unsubscribe;
    }
}
exports.Unsubscribe = Unsubscribe;
