// @signalwire/node/dist/common/src/util/constants/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.STORAGE_PREFIX = '@signalwire:';
exports.ADD = 'add';
exports.REMOVE = 'remove';
exports.SESSION_ID = 'sessId';
var SwEvent;
(function (SwEvent) {
    SwEvent["SocketOpen"] = "signalwire.socket.open";
    SwEvent["SocketClose"] = "signalwire.socket.close";
    SwEvent["SocketError"] = "signalwire.socket.error";
    SwEvent["SocketMessage"] = "signalwire.socket.message";
    SwEvent["SpeedTest"] = "signalwire.internal.speedtest";
    SwEvent["Ready"] = "signalwire.ready";
    SwEvent["Error"] = "signalwire.error";
    SwEvent["Notification"] = "signalwire.notification";
    SwEvent["Messages"] = "signalwire.messages";
    SwEvent["Calls"] = "signalwire.calls";
    SwEvent["MediaError"] = "signalwire.rtc.mediaError";
})(SwEvent = exports.SwEvent || (exports.SwEvent = {}));
var BladeMethod;
(function (BladeMethod) {
    BladeMethod["Broadcast"] = "blade.broadcast";
    BladeMethod["Disconnect"] = "blade.disconnect";
})(BladeMethod = exports.BladeMethod || (exports.BladeMethod = {}));
