// @signalwire/node/dist/common/src/util/constants/relay.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var CallState;
(function (CallState) {
    CallState["None"] = "none";
    CallState["Created"] = "created";
    CallState["Ringing"] = "ringing";
    CallState["Answered"] = "answered";
    CallState["Ending"] = "ending";
    CallState["Ended"] = "ended";
})(CallState = exports.CallState || (exports.CallState = {}));
exports.CALL_STATES = Object.values(CallState);
exports.DEFAULT_CALL_TIMEOUT = 30;
var DisconnectReason;
(function (DisconnectReason) {
    DisconnectReason["Hangup"] = "hangup";
    DisconnectReason["Cancel"] = "cancel";
    DisconnectReason["Busy"] = "busy";
    DisconnectReason["NoAnswer"] = "noAnswer";
    DisconnectReason["Decline"] = "decline";
    DisconnectReason["Error"] = "error";
})(DisconnectReason = exports.DisconnectReason || (exports.DisconnectReason = {}));
var DisconnectSource;
(function (DisconnectSource) {
    DisconnectSource["None"] = "none";
    DisconnectSource["Client"] = "client";
    DisconnectSource["Server"] = "server";
    DisconnectSource["Endpoint"] = "endpoint";
})(DisconnectSource = exports.DisconnectSource || (exports.DisconnectSource = {}));
var CallType;
(function (CallType) {
    CallType["Phone"] = "phone";
    CallType["Sip"] = "sip";
    CallType["WebRTC"] = "webrtc";
})(CallType = exports.CallType || (exports.CallType = {}));
var CallConnectState;
(function (CallConnectState) {
    CallConnectState["Disconnected"] = "disconnected";
    CallConnectState["Connecting"] = "connecting";
    CallConnectState["Connected"] = "connected";
    CallConnectState["Failed"] = "failed";
})(CallConnectState = exports.CallConnectState || (exports.CallConnectState = {}));
var CallNotification;
(function (CallNotification) {
    CallNotification["Dial"] = "calling.call.dial";
    CallNotification["State"] = "calling.call.state";
    CallNotification["Receive"] = "calling.call.receive";
    CallNotification["Connect"] = "calling.call.connect";
    CallNotification["Record"] = "calling.call.record";
    CallNotification["Play"] = "calling.call.play";
    CallNotification["Collect"] = "calling.call.collect";
    CallNotification["Fax"] = "calling.call.fax";
    CallNotification["Detect"] = "calling.call.detect";
    CallNotification["Tap"] = "calling.call.tap";
    CallNotification["SendDigits"] = "calling.call.send_digits";
})(CallNotification = exports.CallNotification || (exports.CallNotification = {}));
var CallMethod;
(function (CallMethod) {
    CallMethod["Answer"] = "calling.answer";
    CallMethod["Begin"] = "calling.begin";
    CallMethod["Connect"] = "calling.connect";
    CallMethod["Dial"] = "calling.dial";
    CallMethod["Disconnect"] = "calling.disconnect";
    CallMethod["End"] = "calling.end";
    CallMethod["Record"] = "calling.record";
    CallMethod["Play"] = "calling.play";
    CallMethod["PlayAndCollect"] = "calling.play_and_collect";
    CallMethod["ReceiveFax"] = "calling.receive_fax";
    CallMethod["SendFax"] = "calling.send_fax";
    CallMethod["Detect"] = "calling.detect";
    CallMethod["Tap"] = "calling.tap";
    CallMethod["SendDigits"] = "calling.send_digits";
})(CallMethod = exports.CallMethod || (exports.CallMethod = {}));
var DialState;
(function (DialState) {
    DialState["Dialing"] = "dialing";
    DialState["Failed"] = "failed";
    DialState["Answered"] = "answered";
})(DialState = exports.DialState || (exports.DialState = {}));
var SendDigitsState;
(function (SendDigitsState) {
    SendDigitsState["Finished"] = "finished";
})(SendDigitsState = exports.SendDigitsState || (exports.SendDigitsState = {}));
var CallPlayState;
(function (CallPlayState) {
    CallPlayState["Playing"] = "playing";
    CallPlayState["Error"] = "error";
    CallPlayState["Finished"] = "finished";
})(CallPlayState = exports.CallPlayState || (exports.CallPlayState = {}));
var CallPlayType;
(function (CallPlayType) {
    CallPlayType["Audio"] = "audio";
    CallPlayType["Silence"] = "silence";
    CallPlayType["TTS"] = "tts";
    CallPlayType["Ringtone"] = "ringtone";
})(CallPlayType = exports.CallPlayType || (exports.CallPlayType = {}));
var CallTapState;
(function (CallTapState) {
    CallTapState["Tapping"] = "tapping";
    CallTapState["Finished"] = "finished";
})(CallTapState = exports.CallTapState || (exports.CallTapState = {}));
var CallFaxState;
(function (CallFaxState) {
    CallFaxState["Page"] = "page";
    CallFaxState["Error"] = "error";
    CallFaxState["Finished"] = "finished";
})(CallFaxState = exports.CallFaxState || (exports.CallFaxState = {}));
var CallPromptState;
(function (CallPromptState) {
    CallPromptState["Error"] = "error";
    CallPromptState["NoInput"] = "no_input";
    CallPromptState["NoMatch"] = "no_match";
    CallPromptState["Digit"] = "digit";
    CallPromptState["Speech"] = "speech";
})(CallPromptState = exports.CallPromptState || (exports.CallPromptState = {}));
var CallRecordState;
(function (CallRecordState) {
    CallRecordState["Recording"] = "recording";
    CallRecordState["NoInput"] = "no_input";
    CallRecordState["Finished"] = "finished";
})(CallRecordState = exports.CallRecordState || (exports.CallRecordState = {}));
var CallDetectState;
(function (CallDetectState) {
    CallDetectState["Error"] = "error";
    CallDetectState["Finished"] = "finished";
    CallDetectState["CED"] = "CED";
    CallDetectState["CNG"] = "CNG";
    CallDetectState["Machine"] = "MACHINE";
    CallDetectState["Human"] = "HUMAN";
    CallDetectState["Unknown"] = "UNKNOWN";
    CallDetectState["Ready"] = "READY";
    CallDetectState["NotReady"] = "NOT_READY";
})(CallDetectState = exports.CallDetectState || (exports.CallDetectState = {}));
var CallDetectType;
(function (CallDetectType) {
    CallDetectType["Fax"] = "fax";
    CallDetectType["Machine"] = "machine";
    CallDetectType["Digit"] = "digit";
})(CallDetectType = exports.CallDetectType || (exports.CallDetectType = {}));
var MessageNotification;
(function (MessageNotification) {
    MessageNotification["State"] = "messaging.state";
    MessageNotification["Receive"] = "messaging.receive";
})(MessageNotification = exports.MessageNotification || (exports.MessageNotification = {}));
