// @signalwire/node/dist/common/src/relay/calling/actions/PlayAction.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const BaseAction_1 = require("./BaseAction");
const results_1 = require("../results");
class PlayAction extends BaseAction_1.BaseAction {
    constructor(component) {
        super(component);
        this.component = component;
    }
    get result() {
        return new results_1.PlayResult(this.component);
    }
    stop() {
        return this.component.stop();
    }
    pause() {
        return __awaiter(this, void 0, void 0, function* () {
            const res = yield this.component.pause();
            return new results_1.PlayPauseResult(res);
        });
    }
    resume() {
        return __awaiter(this, void 0, void 0, function* () {
            const res = yield this.component.resume();
            return new results_1.PlayResumeResult(res);
        });
    }
    volume(value) {
        return __awaiter(this, void 0, void 0, function* () {
            const res = yield this.component.volume(value);
            return new results_1.PlayVolumeResult(res);
        });
    }
}
exports.PlayAction = PlayAction;
