// @signalwire/node/dist/common/src/relay/calling/components/Answer.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const BaseComponent_1 = require("./BaseComponent");
const relay_1 = require("../../../util/constants/relay");
const Event_1 = __importDefault(require("../Event"));
class Answer extends BaseComponent_1.BaseComponent {
    constructor() {
        super(...arguments);
        this.eventType = relay_1.CallNotification.State;
        this.method = relay_1.CallMethod.Answer;
        this.controlId = this.call.tag;
    }
    get payload() {
        return {
            node_id: this.call.nodeId,
            call_id: this.call.id
        };
    }
    notificationHandler(params) {
        const { call_state } = params;
        if (call_state === relay_1.CallState.Answered) {
            this.completed = true;
            this.successful = true;
            this.event = new Event_1.default(call_state, params);
        }
        if (this._hasBlocker() && this._eventsToWait.includes(call_state)) {
            this.blocker.resolve();
        }
    }
}
exports.Answer = Answer;
