// @signalwire/node/dist/common/src/relay/calling/components/BaseFax.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const Controllable_1 = require("./Controllable");
const relay_1 = require("../../../util/constants/relay");
const Event_1 = __importDefault(require("../Event"));
class BaseFax extends Controllable_1.Controllable {
    constructor() {
        super(...arguments);
        this.eventType = relay_1.CallNotification.Fax;
        this.controlId = this.controlId;
    }
    notificationHandler(params) {
        const { fax: { type, params: faxParams } } = params;
        this.state = type;
        this.completed = this.state !== relay_1.CallFaxState.Page;
        if (this.completed) {
            if (faxParams.success) {
                this.successful = true;
                this.direction = faxParams.direction;
                this.identity = faxParams.identity;
                this.remoteIdentity = faxParams.remote_identity;
                this.document = faxParams.document;
                this.pages = faxParams.pages;
            }
            this.event = new Event_1.default(this.state, faxParams);
        }
        if (this._hasBlocker() && this._eventsToWait.includes(this.state)) {
            this.blocker.resolve();
        }
    }
}
exports.BaseFax = BaseFax;
