// @signalwire/node/dist/common/src/relay/calling/components/Connect.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const BaseComponent_1 = require("./BaseComponent");
const relay_1 = require("../../../util/constants/relay");
const Event_1 = __importDefault(require("../Event"));
class Connect extends BaseComponent_1.BaseComponent {
    constructor(call, devices, ringback) {
        super(call);
        this.call = call;
        this.devices = devices;
        this.ringback = ringback;
        this.eventType = relay_1.CallNotification.Connect;
        this.method = relay_1.CallMethod.Connect;
        this.controlId = this.call.tag;
    }
    get payload() {
        const tmp = {
            node_id: this.call.nodeId,
            call_id: this.call.id,
            devices: this.devices
        };
        if (this.ringback) {
            tmp.ringback = this.ringback;
        }
        return tmp;
    }
    notificationHandler(params) {
        this.state = params.connect_state;
        this.completed = this.state !== relay_1.CallConnectState.Connecting;
        if (this.completed) {
            this.successful = this.state === relay_1.CallConnectState.Connected;
            this.event = new Event_1.default(this.state, params);
        }
        if (this._hasBlocker() && this._eventsToWait.includes(this.state)) {
            this.blocker.resolve();
        }
    }
}
exports.Connect = Connect;
