// @signalwire/node/dist/common/src/relay/calling/components/Hangup.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const BaseComponent_1 = require("./BaseComponent");
const relay_1 = require("../../../util/constants/relay");
const Event_1 = __importDefault(require("../Event"));
class Hangup extends BaseComponent_1.BaseComponent {
    constructor(call, reason) {
        super(call);
        this.call = call;
        this.reason = reason;
        this.eventType = relay_1.CallNotification.State;
        this.method = relay_1.CallMethod.End;
        this.controlId = this.call.tag;
    }
    get payload() {
        return {
            node_id: this.call.nodeId,
            call_id: this.call.id,
            reason: this.reason
        };
    }
    notificationHandler(params) {
        const { call_state, end_reason } = params;
        this.state = call_state;
        this.completed = this.state === relay_1.CallState.Ended;
        if (this.completed) {
            this.successful = true;
            this.reason = end_reason;
            this.event = new Event_1.default(this.state, params);
        }
        if (this._hasBlocker() && this._eventsToWait.includes(this.state)) {
            this.reason = end_reason;
            this.blocker.resolve();
        }
    }
}
exports.Hangup = Hangup;
