// @signalwire/node/dist/common/src/relay/calling/components/Record.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const Controllable_1 = require("./Controllable");
const relay_1 = require("../../../util/constants/relay");
const Event_1 = __importDefault(require("../Event"));
class Record extends Controllable_1.Controllable {
    constructor(call, record) {
        super(call);
        this.call = call;
        this.record = record;
        this.eventType = relay_1.CallNotification.Record;
        this.method = relay_1.CallMethod.Record;
        this.controlId = this.controlId;
    }
    get payload() {
        return {
            node_id: this.call.nodeId,
            call_id: this.call.id,
            control_id: this.controlId,
            record: this.record
        };
    }
    notificationHandler(params) {
        const { state, url, duration, size } = params;
        this.state = state;
        this.completed = this.state !== relay_1.CallRecordState.Recording;
        if (this.completed) {
            this.successful = this.state === relay_1.CallRecordState.Finished;
            this.event = new Event_1.default(this.state, params);
            this.url = url;
            this.duration = duration;
            this.size = size;
        }
        if (this._hasBlocker() && this._eventsToWait.includes(this.state)) {
            this.blocker.resolve();
        }
    }
}
exports.Record = Record;
