// @signalwire/node/dist/common/src/relay/calling/components/SendDigits.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const BaseComponent_1 = require("./BaseComponent");
const relay_1 = require("../../../util/constants/relay");
const Event_1 = __importDefault(require("../Event"));
class SendDigits extends BaseComponent_1.BaseComponent {
    constructor(call, digits) {
        super(call);
        this.call = call;
        this.digits = digits;
        this.eventType = relay_1.CallNotification.SendDigits;
        this.method = relay_1.CallMethod.SendDigits;
        this.controlId = this.controlId;
    }
    get payload() {
        return {
            node_id: this.call.nodeId,
            call_id: this.call.id,
            control_id: this.controlId,
            digits: this.digits
        };
    }
    notificationHandler(params) {
        this.state = params.state;
        this.completed = this.state === relay_1.SendDigitsState.Finished;
        this.successful = this.completed;
        this.event = new Event_1.default(this.state, params);
        if (this._hasBlocker() && this._eventsToWait.includes(this.state)) {
            this.blocker.resolve();
        }
    }
}
exports.SendDigits = SendDigits;
