// @signalwire/node/dist/common/src/relay/tasking/Task.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const https_1 = require("https");
class Task {
    constructor(project, token) {
        this.project = project;
        this.token = token;
        this.host = 'relay.signalwire.com';
        if (!project || !token) {
            throw new Error('Invalid options: project and token required!');
        }
    }
    deliver(context, message) {
        return new Promise((resolve, reject) => {
            const data = JSON.stringify({ context, message });
            const options = {
                host: this.host,
                port: 443,
                method: 'POST',
                path: '/api/relay/rest/tasks',
                headers: {
                    'Authorization': 'Basic ' + Buffer.from(`${this.project}:${this.token}`).toString('base64'),
                    'Content-Type': 'application/json',
                    'Content-Length': data.length
                }
            };
            const req = https_1.request(options, ({ statusCode }) => {
                statusCode === 204 ? resolve() : reject();
            });
            req.on('error', reject);
            req.write(data);
            req.end();
        });
    }
}
exports.default = Task;
